File: src/main/Python/nowcoder/S4_ReConstructBinaryTree.py
from PrintTreeLayer import PrintTreeLayer
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
class Solution:
    def reConstructBinaryTree(self, pre, tin):
        # write code here
        if len(pre) == 0:
            return None
        if len(pre) == 1:
            return TreeNode(pre[0])
        else:
            root = TreeNode(pre[0])
            root.left = self.reConstructBinaryTree(pre[1:tin.index(pre[0])+1], tin[:tin.index(pre[0])])
            root.right = self.reConstructBinaryTree(pre[tin.index(pre[0])+1:], tin[tin.index(pre[0])+1:])
        return root

if __name__ =='__main__':
    pre = [3, 9, 20, 15, 7]
    tin = [9, 3, 15, 20, 7]
    test = Solution()
    result = test.reConstructBinaryTree(pre, tin)
    prt = PrintTreeLayer()
    prt.printTreeLayer(result)

File: src/main/java/nowcoder/S4_ReConstructBinaryTree.java
package nowcoder;

/**
 * @author LiuZhiguo
 * @date 2019/12/13 19:20
 */
public class S4_ReConstructBinaryTree {
    public TreeNode reConstructBinaryTree(int[] pre, int[] in){
        TreeNode root = reConstruct(pre, 0, pre.length-1, in, 0, in.length-1);
        return root;
    }
    private TreeNode reConstruct(int[] pre, int preStart, int preEnd, int[] in, int inStart, int inEnd){
        if (preStart > preEnd || inStart > inEnd)
            return null;
        TreeNode root = new TreeNode(pre[preStart]); //前序第一个即是根节点
        for (int i=inStart;i<=inEnd;i++){
            if (in[i] == pre[preStart]){
                root.left = reConstruct(pre, preStart+1, preStart+i-inStart, in, inStart,i-1);
                root.right = reConstruct(pre, preStart+i-inStart+1, preEnd, in, i+1, inEnd);
                break;
            }
        }
        return root;
    }
    public static void main(String[] args){
        S4_ReConstructBinaryTree s4 = new S4_ReConstructBinaryTree();
        int[] pre = {3, 9, 20, 15, 7};
        int[] in = {9, 3, 15, 20, 7};
        TreeNode root = s4.reConstructBinaryTree(pre, in);
        PrintTreeLayer p = new PrintTreeLayer();
        p.printTreeLayer(root);
    }
}
